### ISSUE TITLE
Race condition in 'mixer build bundles'

### ISSUE THREAD

**Issue by cmarcelo** [clearlinux/mixer-tools#348]
Just creating a mix with init, then running repeatedly mixer build bundles was sufficient to trigger. Sometimes I get it in the first run.

See the "fatal error: concurrent map writes" snippet below.

```
Resolving packages using 4 workers
processing os-core
processing bootloader
processing os-core-update
processing kernel-native
... done with os-core
... done with os-core-update
fatal error: ... done with bootloader
concurrent map writes
```

and the offending goroutine:

```
goroutine 37 [running]:
runtime.throw(0x868c9a, 0x15)
	/usr/lib/golang/src/runtime/panic.go:616 +0x81 fp=0xc4202dac58 sp=0xc4202dac38 pc=0x42b691
runtime.mapassign_faststr(0x7f14c0, 0xc420169740, 0xc420538fab, 0xd, 0x0)
	/usr/lib/golang/src/runtime/hashmap_fast.go:779 +0x3ce fp=0xc4202dacc8 sp=0xc4202dac58 pc=0x40d82e
github.com/clearlinux/mixer-tools/builder.resolvePackages.func1()
	/builddir/build/BUILD/mixer-tools-4.6.5/.gopath/src/github.com/clearlinux/mixer-tools/builder/bundles.go:378 +0x41b fp=0xc4202dafe0 sp=0xc4202dacc8 pc=0x750ddb
runtime.goexit()
	/usr/lib/golang/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc4202dafe8 sp=0xc4202dafe0 pc=0x457f01
created by github.com/clearlinux/mixer-tools/builder.resolvePackages
	/builddir/build/BUILD/mixer-tools-4.6.5/.gopath/src/github.com/clearlinux/mixer-tools/builder/bundles.go:392 +0x2de
```

Full trace:

```
$ mixer build bundles
Building bundles...
=> BUILD BUNDLES
Wiping away previous version 10...
Creating bundles for version 10 based on Clear Linux 23000
Preparing new /home/c/src/mix2/update/image/10
  and dnf config: /home/c/src/mix2/.yum-mix.conf
Packager command-line: dnf --config=/home/c/src/mix2/.yum-mix.conf -y --releasever=23000
Resolving packages using 4 workers
processing os-core
processing bootloader
processing os-core-update
processing kernel-native
... done with os-core
... done with os-core-update
fatal error: ... done with bootloader
concurrent map writes

goroutine 37 [running]:
runtime.throw(0x868c9a, 0x15)
	/usr/lib/golang/src/runtime/panic.go:616 +0x81 fp=0xc4202dac58 sp=0xc4202dac38 pc=0x42b691
runtime.mapassign_faststr(0x7f14c0, 0xc420169740, 0xc420538fab, 0xd, 0x0)
	/usr/lib/golang/src/runtime/hashmap_fast.go:779 +0x3ce fp=0xc4202dacc8 sp=0xc4202dac58 pc=0x40d82e
github.com/clearlinux/mixer-tools/builder.resolvePackages.func1()
	/builddir/build/BUILD/mixer-tools-4.6.5/.gopath/src/github.com/clearlinux/mixer-tools/builder/bundles.go:378 +0x41b fp=0xc4202dafe0 sp=0xc4202dacc8 pc=0x750ddb
runtime.goexit()
	/usr/lib/golang/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc4202dafe8 sp=0xc4202dafe0 pc=0x457f01
created by github.com/clearlinux/mixer-tools/builder.resolvePackages
	/builddir/build/BUILD/mixer-tools-4.6.5/.gopath/src/github.com/clearlinux/mixer-tools/builder/bundles.go:392 +0x2de

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0xc420348d4c)
	/usr/lib/golang/src/runtime/sema.go:56 +0x39
sync.(*WaitGroup).Wait(0xc420348d40)
	/usr/lib/golang/src/sync/waitgroup.go:129 +0x72
github.com/clearlinux/mixer-tools/builder.resolvePackages(0x4, 0xc420168960, 0xc42012c3c0, 0x4, 0x4, 0xc4201a11d0, 0x29, 0x0)
	/builddir/build/BUILD/mixer-tools-4.6.5/.gopath/src/github.com/clearlinux/mixer-tools/builder/bundles.go:401 +0x3c4
github.com/clearlinux/mixer-tools/builder.(*Builder).buildBundles(0xc420156000, 0xc420168960, 0x0, 0x0)
	/builddir/build/BUILD/mixer-tools-4.6.5/.gopath/src/github.com/clearlinux/mixer-tools/builder/bundles.go:740 +0x1040
github.com/clearlinux/mixer-tools/builder.(*Builder).BuildBundles(0xc420156000, 0x0, 0x0, 0xc4202fbd01, 0x0, 0x0)
	/builddir/build/BUILD/mixer-tools-4.6.5/.gopath/src/github.com/clearlinux/mixer-tools/builder/builder.go:1450 +0x2d6
github.com/clearlinux/mixer-tools/mixer/cmd.buildBundles(0xc420156000, 0x0, 0xc420156000, 0x0)
	/builddir/build/BUILD/mixer-tools-4.6.5/.gopath/src/github.com/clearlinux/mixer-tools/mixer/cmd/build.go:91 +0x1ae
github.com/clearlinux/mixer-tools/mixer/cmd.glob..func1(0xaaeb00, 0xad51d8, 0x0, 0x0)
	/builddir/build/BUILD/mixer-tools-4.6.5/.gopath/src/github.com/clearlinux/mixer-tools/mixer/cmd/build.go:110 +0xba
github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra.(*Command).execute(0xaaeb00, 0xad51d8, 0x0, 0x0, 0xaaeb00, 0xad51d8)
	/builddir/build/BUILD/mixer-tools-4.6.5/.gopath/src/github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra/command.go:702 +0x2c6
github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xab2020, 0x8643fa, 0xa, 0x86ba99)
	/builddir/build/BUILD/mixer-tools-4.6.5/.gopath/src/github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra/command.go:783 +0x2e4
github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra.(*Command).Execute(0xab2020, 0xa, 0x8726e0)
	/builddir/build/BUILD/mixer-tools-4.6.5/.gopath/src/github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra/command.go:736 +0x2b
github.com/clearlinux/mixer-tools/mixer/cmd.Execute()
	/builddir/build/BUILD/mixer-tools-4.6.5/.gopath/src/github.com/clearlinux/mixer-tools/mixer/cmd/root.go:188 +0x2d
main.main()
	/builddir/build/BUILD/mixer-tools-4.6.5/.gopath/src/github.com/clearlinux/mixer-tools/mixer/main.go:29 +0x5e

goroutine 14 [IO wait]:
internal/poll.runtime_pollWait(0x7f56a805ee30, 0x72, 0xc420051760)
	/usr/lib/golang/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc420278018, 0x72, 0xffffffffffffff00, 0x8b9640, 0xa7a510)
	/usr/lib/golang/src/internal/poll/fd_poll_runtime.go:85 +0x9b
internal/poll.(*pollDesc).waitRead(0xc420278018, 0xc42028a400, 0x400, 0x400)
	/usr/lib/golang/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc420278000, 0xc42028a400, 0x400, 0x400, 0x0, 0x0, 0x0)
	/usr/lib/golang/src/internal/poll/fd_unix.go:157 +0x17d
net.(*netFD).Read(0xc420278000, 0xc42028a400, 0x400, 0x400, 0xc4200518e0, 0xc4200518a0, 0x411a36)
	/usr/lib/golang/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc420130038, 0xc42028a400, 0x400, 0x400, 0x0, 0x0, 0x0)
	/usr/lib/golang/src/net/net.go:176 +0x6a
crypto/tls.(*block).readFromUntil(0xc42012a480, 0x7f56a0f7a000, 0xc420130038, 0x5, 0xc420130038, 0x1)
	/usr/lib/golang/src/crypto/tls/conn.go:493 +0x96
crypto/tls.(*Conn).readRecord(0xc420284000, 0x884217, 0xc420284120, 0x454160)
	/usr/lib/golang/src/crypto/tls/conn.go:595 +0xe0
crypto/tls.(*Conn).Read(0xc420284000, 0xc4202c8000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/lib/golang/src/crypto/tls/conn.go:1156 +0x100
net/http.(*persistConn).Read(0xc4201d07e0, 0xc4202c8000, 0x1000, 0x1000, 0xc420051b98, 0x4051c5, 0xc4204280c0)
	/usr/lib/golang/src/net/http/transport.go:1453 +0x136
bufio.(*Reader).fill(0xc420128de0)
	/usr/lib/golang/src/bufio/bufio.go:100 +0x11e
bufio.(*Reader).Peek(0xc420128de0, 0x1, 0x0, 0x0, 0x0, 0xc420015f20, 0x0)
	/usr/lib/golang/src/bufio/bufio.go:132 +0x3a
net/http.(*persistConn).readLoop(0xc4201d07e0)
	/usr/lib/golang/src/net/http/transport.go:1601 +0x185
created by net/http.(*Transport).dialConn
	/usr/lib/golang/src/net/http/transport.go:1237 +0x95a

goroutine 15 [select]:
net/http.(*persistConn).writeLoop(0xc4201d07e0)
	/usr/lib/golang/src/net/http/transport.go:1822 +0x14b
created by net/http.(*Transport).dialConn
	/usr/lib/golang/src/net/http/transport.go:1238 +0x97f
```

**Comment by matthewrsj** [comment]
> running repeatedly mixer build bundles

What do you mean by this? Running it until completion then immediately running it again or running it concurrently with itself in the same directory?

**Comment by cmarcelo** [comment]
Run and wait completion. Then run again the same command.

I’ve cleared the whole directory and restarted with init a few times to make sure.

**Comment by matthewrsj** [comment]
A race certainly exists
```bash
$ sudo -E go run -race ../mixer/main.go build bundles --native
Generating certificate for signature validation...
Building bundles...
=> BUILD BUNDLES
Creating bundles for version 10 based on Clear Linux 24440
Preparing new /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/mix/update/image/10
  and dnf config: /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/mix/.yum-mix.conf
Packager command-line: dnf --config=/home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/mix/.yum-mix.conf -y --releasever=24440
Resolving packages using 12 workers
processing os-core-update
processing bootloader
processing os-core
processing kernel-native
... done with kernel-native
... done with bootloader
... done with os-core-update
... done with os-core
Resolving files using 12 workers
processing os-core
processing kernel-native
processing os-core-update
processing bootloader
Bundle bootloader	329 files
Bundle os-core	2939 files
==================
WARNING: DATA RACE
Write at 0x00c420328100 by goroutine 54:
  github.com/clearlinux/mixer-tools/builder.resolveFiles.func1()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/builder/bundles.go:235 +0x245

Previous write at 0x00c420328100 by goroutine 53:
  github.com/clearlinux/mixer-tools/builder.resolveFiles.func1()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/builder/bundles.go:235 +0x245

Goroutine 54 (running) created at:
  github.com/clearlinux/mixer-tools/builder.resolveFiles()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/builder/bundles.go:249 +0x395
  github.com/clearlinux/mixer-tools/builder.(*Builder).buildBundles()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/builder/bundles.go:742 +0x17f0
  github.com/clearlinux/mixer-tools/builder.(*Builder).BuildBundles()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/builder/builder.go:1465 +0x3c1
  github.com/clearlinux/mixer-tools/mixer/cmd.buildBundles()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/mixer/cmd/build.go:87 +0x122
  github.com/clearlinux/mixer-tools/mixer/cmd.glob..func1()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/mixer/cmd/build.go:111 +0xc0
  github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra.(*Command).execute()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra/command.go:702 +0x881
  github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra.(*Command).ExecuteC()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra/command.go:783 +0x414
  github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra.(*Command).Execute()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra/command.go:736 +0x38
  github.com/clearlinux/mixer-tools/mixer/cmd.Execute()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/mixer/cmd/root.go:207 +0x4a
  main.main()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/mixer/main.go:22 +0x2f

Goroutine 53 (finished) created at:
  github.com/clearlinux/mixer-tools/builder.resolveFiles()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/builder/bundles.go:249 +0x395
  github.com/clearlinux/mixer-tools/builder.(*Builder).buildBundles()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/builder/bundles.go:742 +0x17f0
  github.com/clearlinux/mixer-tools/builder.(*Builder).BuildBundles()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/builder/builder.go:1465 +0x3c1
  github.com/clearlinux/mixer-tools/mixer/cmd.buildBundles()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/mixer/cmd/build.go:87 +0x122
  github.com/clearlinux/mixer-tools/mixer/cmd.glob..func1()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/mixer/cmd/build.go:111 +0xc0
  github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra.(*Command).execute()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra/command.go:702 +0x881
  github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra.(*Command).ExecuteC()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra/command.go:783 +0x414
  github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra.(*Command).Execute()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/vendor/github.com/spf13/cobra/command.go:736 +0x38
  github.com/clearlinux/mixer-tools/mixer/cmd.Execute()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/mixer/cmd/root.go:207 +0x4a
  main.main()
      /home/mrsj/clearlinux/projects/go/src/github.com/clearlinux/mixer-tools/mixer/main.go:22 +0x2f
==================
Bundle os-core-update	3463 files
Bundle kernel-native	4787 files
Cleaning DNF cache before full install
Installing all bundles to full chroot
[1/4] kernel-native
[2/4] bootloader
[3/4] os-core
... building special os-core content
[4/4] os-core-update
... Adding swupd default values to os-core-update bundle

TIMINGS
  BUILD BUNDLES 22.956s
TOTAL: 22.956s
Found 1 data race(s)
exit status 66
```

**Comment by matthewrsj** [comment]
Fixed by #403
